akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala

```diff
@@ -268,4 +277,44 @@ import akka.util.Timeout
    * INTERNAL API
    */
   @InternalApi private[akka] def messageAdapters: List[(Class[_], Any => T)] = _messageAdapters
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def setCurrentActorThread(): Unit = {
+    _currentActorThread match {
+      case OptionVal.None =>
+        _currentActorThread = OptionVal.Some(Thread.currentThread())
+      case OptionVal.Some(t) =>
+        throw new IllegalStateException(
+          s"Invalid access by thread from the outside of $self. " +
+          s"Current message is processed by $t, but also accessed from from ${Thread.currentThread()}.")
+    }
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def clearCurrentActorThread(): Unit = {
+    _currentActorThread = OptionVal.None
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def checkCurrentActorThread(operationName: String): Unit = {
+    val callerThread = Thread.currentThread()
+    _currentActorThread match {
+      case OptionVal.Some(t) =>
+        if (callerThread ne t) {
```

Comment:
Possible flatten to three cases with ```case OptionVal(`callerThread`)``` as first case

Reply:
That would use `equals` instead of `eq`. Not sure if that matters for Thread. Also not sure if that would influence the pattern match and zero allocation of OptionVal. For these reasons I prefer to stick the `if` to be sure what is going on.